crates/revm/src/evm_impl.rs

```diff
@@ -85,10 +88,35 @@ impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> EVMImpl<'a, GSPEC, DB,
 impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> Transact<DB::Error>
     for EVMImpl<'a, GSPEC, DB, INSPECT>
 {
-    fn transact(&mut self) -> EVMResult<DB::Error> {
-        self.env().validate_block_env::<GSPEC, DB::Error>()?;
-        self.env().validate_tx::<GSPEC>()?;
+    fn preverify_transaction(&mut self) -> Result<(), EVMError<DB::Error>> {
+        let env = self.env();
+
+        env.validate_block_env::<GSPEC, DB::Error>()?;
+        env.validate_tx::<GSPEC>()?;
+
+        let tx_caller = env.tx.caller;
+        let tx_data = &env.tx.data;
+        let tx_is_create = env.tx.transact_to.is_create();
 
+        let initial_gas_spend = initial_tx_gas::<GSPEC>(tx_data, tx_is_create, &env.tx.access_list);
+
+        // Additonal check to see if limit is big enought to cover initial gas.
+        if env.tx.gas_limit < initial_gas_spend {
+            return Err(InvalidTransaction::CallGasCostMoreThanGasLimit.into());
+        }
+
+        // load acc
+        let journal = &mut self.data.journaled_state;
+        let (caller_account, _) = journal
+            .load_account(tx_caller, self.data.db)
+            .map_err(EVMError::Database)?;
+
+        self.data.env.validate_tx_against_state(caller_account)?;
+
+        Ok(())
+    }
+
+    fn transact(&mut self) -> EVMResult<DB::Error> {
```

Comment:
It will be a foot gun if we silently remove some checks. 
We should probably add an additional function `transact_preverified` to make it explicit.

Reply:
You're right! Sorry for the dumb take. Now I added the `transact_preverified` function in the `Transact` trait